### PR TITLE
Add railroad diagrams

### PR DESCRIPTION
**Pull request by konnov**
@pdini mentioned that it would be good to have a look at the syntax. This PR contains a railroad diagram of the current Quint syntax. Unfortunately, the best diagrams are generated via the website http://bottlecaps.de/rr/ui. Hence, it have to be manually updated.

When I looked at the generated diagrams, I saw a way to trim the current grammar by removing some unnecessary tokens. So this PR also makes the grammar even smaller.

- [x] Entries added to the respective `CHANGELOG.md` for any new functionality

### CODE REVIEW

**Comment by mingodad** [comment]
It can be generated offline too by downloading (https://www.bottlecaps.de/rr/download/rr-2.0-java11.zip) and executing it :
```
java -jar rr.war ebnf-file
```

**Comment by konnov** [comment]
> It can be generated offline too by downloading (https://www.bottlecaps.de/rr/download/rr-2.0-java11.zip) and executing it :
> 
> ```
> java -jar rr.war ebnf-file
> ```

Oh. I did not know that. Thanks a lot!

File: quint/src/generated/Quint.g4 [hunk 1]
```diff
@@ -167,25 +171,19 @@ identOrStar :   '*' | IDENTIFIER
...
 
-argList :      expr (',' expr)*
+argList :      '(' expr? (',' expr)* ')'
         ;
 
 // operators in the normal call may use a few reserved names,
 // which are not recognized as identifiers.
 normalCallName :   IDENTIFIER
-        |       op=(AND | OR | IFF | IMPLIES | SET | LIST | MAP)
+        |       op=('and' | 'or' | 'iff' | 'implies' | SET | LIST | MAP)
```
Comment (shonfeder) [inline quint/src/generated/Quint.g4:180]: See, now we have duplicate stings on at least 4 different places, and each one is an opportunity for a typo.

File: quint/src/generated/Quint.g4 [hunk 2]
```diff
@@ -93,36 +97,36 @@ row : | (IDENTIFIER ':' type ',')* ((IDENTIFIER ':' type) (',' | '|' (IDENTIFIER
 // check the precedence table in Java:
 // https://www.cs.bilkent.edu.tr/~guvenir/courses/CS101/op_precedence.html
 expr:           // apply a built-in operator via the dot notation
-                expr '.' nameAfterDot (LPAREN argList? RPAREN)?     # dotCall
+                expr '.' nameAfterDot argList?                      # dotCall
         |       lambda                                              # lambdaCons
                 // Call a user-defined operator or a built-in operator.
                 // The operator has at least one argument (otherwise, it's a 'val').
-        |       normalCallName '(' argList? ')'                     # operApp
+        |       normalCallName argList                              # operApp
```
Comment (shonfeder) [inline quint/src/generated/Quint.g4:104]: This is the only change in the grammar which I can easily see as a simplification. Could you explain a bit the motive for the other changes?

**Review by shonfeder — COMMENTED** [review]
I've a few points of confusion, but the diagrams look OK to me, assuming we have a clear story on how we can keep them up to date.

**Comment by mingodad** [comment]
Using this online tool https://www.bottlecaps.de/convert/ with https://github.com/informalsystems/quint/blob/main/quint/src/generated/Quint.g4 .

Copy and paste the `EBNF`  shown bellow on https://www.bottlecaps.de/rr/ui on the tab `Edit Grammar` the click on the tab `View Diagram` to see/download a navigable railroad diagram .

```
/* converted on Tue Feb 21, 2023, 22:01 (UTC+01) by antlr_4-to-w3c v0.62 which is Copyright (c) 2011-2022 by Gunther Rademacher <grd@gmx.net> */

modules  ::= module+ EOF
module   ::= 'module' IDENTIFIER '{' ( docLines unit )* '}'
docLines ::= DOCCOMMENT*
unit     ::= ( 'const' | 'var' ) IDENTIFIER ':' type
           | 'assume' identOrHole '=' expr
           | operDef
           | module
           | instanceMod
           | 'type' IDENTIFIER ( '=' type )?
           | 'import' path '.' identOrStar
operDef  ::= qualifier normalCallName ( '(' ( ( IDENTIFIER ( ',' IDENTIFIER )* )? ')' ( ':' type )? | IDENTIFIER ':' type ( ',' IDENTIFIER ':' type )* ')' ':' type ) | ':' type )? ( '=' expr )? ';'?
qualifier
         ::= 'pure'? ( 'val' | 'def' )
           | 'action'
           | 'run'
           | 'temporal'
instanceMod
         ::= 'module' IDENTIFIER '=' IDENTIFIER '(' ( '*' | IDENTIFIER '=' expr ( ',' IDENTIFIER '=' expr )* ( ',' '*' )? ) ')'
type     ::= type ( '->' | '=>' ) type
           | '(' ( ( type ( ',' type )* )? ','? ')' '=>' type | type ( ( ',' type )+ ','? )? ')' )
           | ( 'Set' | 'List' ) '[' type ']'
           | '{' row '}'
           | typeUnionRecOne+
           | 'int'
           | 'str'
           | 'bool'
           | IDENTIFIER
typeUnionRecOne
         ::= '|' '{' IDENTIFIER ':' STRING ( ',' row )? ','? '}'
row      ::= ( ( IDENTIFIER ':' type ',' )* ( IDENTIFIER ':' type ( ',' | '|' IDENTIFIER )? )? | '|' IDENTIFIER )?
expr     ::= expr ( '.' nameAfterDot ( '(' argList? ')' )? | '[' expr ']' | ( '^' | '*' | '/' | '%' | '+' | '-' | '>' | '<' | '>=' | '<=' | '!=' | '==' | '=' | 'and' | 'or' | 'iff' | 'implies' | '->' ) expr | 'match' ( '|' STRING ':' identOrHole '=>' expr )+ )
           | lambda
           | ( normalCallName '(' argList? | '(' expr ( ( ',' expr )+ ','? )? ) ')'
           | ( '-' | 'if' '(' expr ')' expr 'else' | operDef | 'nondet' IDENTIFIER ( ':' type )? '=' expr ';'? ) expr
           | IDENTIFIER ( "'" '=' expr )?
           | ( ( 'and' | 'or' | 'all' | 'any' ) '{' expr ( ',' expr )* ','? | '{' ( IDENTIFIER ':' expr ( ',' IDENTIFIER ':' expr )* ','? | expr ) ) '}'
           | INT
           | BOOL
           | STRING
           | '[' ( expr ( ',' expr )* )? ','? ']'
unitOrExpr
         ::= unit
           | expr
lambda   ::= ( identOrHole | '(' identOrHole ( ',' identOrHole )* ')' ) '=>' expr
identOrHole
         ::= '_'
           | IDENTIFIER
identOrStar
         ::= '*'
           | IDENTIFIER
path     ::= IDENTIFIER ( '.' IDENTIFIER )*
argList  ::= expr ( ',' expr )*
normalCallName
         ::= IDENTIFIER
           | 'and'
           | 'or'
           | 'iff'
           | 'implies'
           | 'Set'
           | 'List'
           | 'Map'
nameAfterDot
         ::= IDENTIFIER
           | 'and'
           | 'or'
           | 'iff'
           | 'implies'
operator ::= 'and'
           | 'or'
           | 'iff'
           | 'implies'
           | '>'
           | '<'
           | '>='
           | '<='
           | '!='
           | '=='
           | '*'
           | '/'
           | '%'
           | '+'
           | '-'
           | '^'
literal  ::= STRING
           | BOOL
           | INT
_        ::= LINE_COMMENT
           | COMMENT
           | WS
          /* ws: definition */

<?TOKENS?>

STRING?  ::= '"' .* '"'
BOOL     ::= 'false'
           | 'true'
INT      ::= '0'
           | [1-9] ( '_'? [0-9] )*
           | '0x' [0-9a-fA-F] ( '_'? [0-9a-fA-F] )*
IDENTIFIER
         ::= SIMPLE_IDENTIFIER ( '::' SIMPLE_IDENTIFIER )*
SIMPLE_IDENTIFIER
         ::= ( [a-zA-Z] | '_' [a-zA-Z0-9_] ) [a-zA-Z0-9_]*
DOCCOMMENT?
         ::= '///' .* #xA
LINE_COMMENT?
         ::= '//' .* #xA
COMMENT? ::= '/*' .* '*/'
WS       ::= [ #x9#xd#xa]+
EOF      ::= $
```

File: doc/railroad.md [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+referenced by:
+
+* argList
+* expr
+* instanceMod
+* lambda
+* operDef
+* unit
+* unitOrExpr
```
[the remark this answers is not in the document]
Comment (konnov) [reply doc/railroad.md:97]: This file is auto-generated, so we do not have to maintain it by hand.

File: doc/railroad.md [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+referenced by:
+
+* expr
+
+**literal:**
+
+![literal](diagram/literal.svg)
+
+**_:**
```
[the remark this answers is not in the document]
Comment (konnov) [reply doc/railroad.md:166]: That's the underscore in the lambda operator. Also auto-generated.

File: quint/src/generated/Quint.g4 [hunk 5]
```diff
@@ -93,36 +97,36 @@ row : | (IDENTIFIER ':' type ',')* ((IDENTIFIER ':' type) (',' | '|' (IDENTIFIER
...
+        |       expr 'and' expr                                     # and
         |       'or'  '{' expr (',' expr)* ','? '}'                 # orExpr
-        |       expr OR expr                                        # or
-        |       expr IFF expr                                       # iff
-        |       expr IMPLIES expr                                   # implies
-        |       expr MATCH
+        |       expr 'or' expr                                      # or
+        |       expr 'iff' expr                                     # iff
+        |       expr 'implies' expr                                 # implies
+        |       expr 'match'
```
[the remark this answers is not in the document]
Comment (konnov) [reply quint/src/generated/Quint.g4:129]: I will bring them back as tokens. It does not improve readability of the grammar though. If we have enough tests, a typo in any of those will pop up immediately.

[the remark this answers is not in the document]
Comment (shonfeder) [reply quint/src/generated/Quint.g4:129]: If indeed :) but why test when can make the class of errors impossible? I can understand the readability improvement view (tho it is not harder for me to read `AND` than `'and'` ;) In addition to reducing error surface, the current approach also helps with maintainability I think?

File: doc/railroad.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (shonfeder) [reply doc/railroad.md:97]: Oh, nice!

If would be helpful if the file head a header indicating it was generated, but lacking that perhaps we should add a note in the readme to explain how this is generated?

File: doc/railroad.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (shonfeder) [reply doc/railroad.md:166]: I see. Thanks for explaining!

**Comment by konnov** [comment]
It looks like we addressed the concern with the [cheatsheet](https://github.com/informalsystems/quint/blob/main/doc/quint-cheatsheet.pdf). Given that this is a large PR, let's close it and revisit at some point later, if needed.

**Comment by mingodad** [comment]
Here is another `EBNF` generated offline with https://github.com/GuntherRademacher/ebnf-convert from the same autor of https://github.com/GuntherRademacher/rr to generate navigable railroad diagram.

Notice that on https://github.com/informalsystems/quint/blob/63f073093dd0b0c34d0ff44d0a40226d81dca22b/quint/src/generated/Quint.g4#L309-L311
**Both `LOW_ID` and `CAP_ID` have `[_][a-zA-Z0-9_]+` which is ambiguous because the lexer only can match one of then not both.**
```
/* converted on Wed Jul 31, 2024, 12:07 (UTC+02) by antlr_4-to-w3c v0.68-SNAPSHOT which is Copyright (c) 2011-2024 by Gunther Rademacher <grd@gmx.net> */

modules  ::= module+ EOF
module   ::= DOCCOMMENT* 'module' qualId '{' documentedDeclaration* '}'
documentedDeclaration
         ::= DOCCOMMENT* declaration
declaration
         ::= ( 'const' | 'var' ) qualId ':' type
           | 'assume' identOrHole '=' expr
           | instanceMod
           | operDef
           | typeDef
           | importMod
           | exportMod
operDef  ::= qualifier normalCallName ( '(' annotatedParameter ( ',' annotatedParameter )* ')' ':' type | ( '(' ( parameter ( ',' parameter )* )? ')' )? ( ':' type )? ) ( '=' expr )? ';'?
typeDef  ::= 'type' ( qualId | typeDefHead '=' ( type | sumTypeDefinition ) )
typeDefHead
         ::= qualId ( '[' LOW_ID ( ',' LOW_ID )* ']' )?
sumTypeDefinition
         ::= '|'? typeSumVariant ( '|' typeSumVariant )*
typeSumVariant
         ::= simpleId ( '(' type ')' )?
qualifier
         ::= 'pure'? ( 'val' | 'def' )
           | 'action'
           | 'run'
           | 'temporal'
           | 'nondet'
importMod
         ::= 'import' name ( '.' identOrStar | 'as' name )? ( 'from' fromSource )?
exportMod
         ::= 'export' name ( '.' identOrStar | 'as' name )?
instanceMod
         ::= 'import' moduleName '(' name '=' expr ( ',' name '=' expr )* ')' ( '.' '*' | 'as' qualifiedName ) ( 'from' fromSource )?
moduleName
         ::= qualId
name     ::= qualId
qualifiedName
         ::= qualId
fromSource
         ::= STRING
type     ::= type ( '->' | '=>' ) type
           | '(' ( ( type ( ',' type )* )? ','? ')' '=>' type | ( type ( ( ',' type )+ ','? )? )? ')' )
           | ( 'Set' | 'List' ) '[' type ']'
           | '{' row? '}'
           | 'int'
           | 'str'
           | 'bool'
           | typeVar
           | qualId ( '[' type ( ',' type )* ']' )?
typeVar  ::= LOW_ID
row      ::= rowLabel ':' type ( ',' rowLabel ':' type )* ( ',' | '|' identifier )?
           | '|' identifier
rowLabel ::= simpleId
expr     ::= expr ( '.' nameAfterDot ( '(' argList? ')' )? | '[' expr ']' | ( '^' | '*' | '/' | '%' | '+' | '-' | '>' | '<' | '>=' | '<=' | '!=' | '==' | '=' | 'and' | 'or' | 'iff' | 'implies' | '->' ) expr )
           | lambda
           | ( normalCallName '(' argList? | '(' ( expr ( ( ',' expr )+ ','? )? )? ) ')'
           | ( '-' | 'if' '(' expr ')' expr 'else' | operDef ) expr
           | qualId ( "'" '=' expr )?
           | ( ( 'and' | 'or' | 'all' | 'any' ) '{' expr ( ',' expr )* ','? | '{' ( recElem ( ',' recElem )* ','? | expr ) ) '}'
           | matchSumExpr
           | INT
           | BOOL
           | STRING
           | '[' ( expr ( ',' expr )* )? ','? ']'
matchSumExpr
         ::= 'match' expr '{' '|'? matchSumCase ( '|' matchSumCase )* '}'
matchSumCase
         ::= ( matchSumVariant | '_' ) '=>' expr
matchSumVariant
         ::= simpleId ( '(' ( simpleId | '_' ) ')' )?
declarationOrExpr
         ::= ( declaration | expr | DOCCOMMENT )? EOF
lambda   ::= lambdaUnsugared
           | lambdaTupleSugar
lambdaUnsugared
         ::= ( parameter | '(' parameter ( ',' parameter )* ')' ) '=>' expr
lambdaTupleSugar
         ::= '(' '(' parameter ( ',' parameter )+ ')' ')' '=>' expr
identOrHole
         ::= '_'
           | qualId
parameter
         ::= identOrHole
annotatedParameter
         ::= identOrHole ':' type
identOrStar
         ::= '*'
           | qualId
argList  ::= expr ( ',' expr )*
recElem  ::= ( simpleId ':' | '...' ) expr
normalCallName
         ::= qualId
           | 'and'
           | 'or'
           | 'iff'
           | 'implies'
           | 'Set'
           | 'List'
           | 'Map'
nameAfterDot
         ::= qualId
           | 'and'
           | 'or'
           | 'iff'
           | 'implies'
operator ::= 'and'
           | 'or'
           | 'iff'
           | 'implies'
           | '>'
           | '<'
           | '>='
           | '<='
           | '!='
           | '=='
           | '*'
           | '/'
           | '%'
           | '+'
           | '-'
           | '^'
literal  ::= STRING
           | BOOL
           | INT
qualId   ::= identifier ( '::' identifier )*
simpleId ::= identifier
           | qualId
identifier
         ::= LOW_ID
           | CAP_ID
_        ::= LINE_COMMENT
           | COMMENT
           | WS
          /* ws: definition */

<?TOKENS?>

STRING?  ::= '"' .* '"'
BOOL     ::= 'false'
           | 'true'
INT      ::= '0'
           | [1-9] ( '_'? [0-9] )*
           | '0x' [0-9a-fA-F] ( '_'? [0-9a-fA-F] )*
LOW_ID   ::= ( [a-z] | '_' [a-zA-Z0-9_] ) [a-zA-Z0-9_]*
CAP_ID   ::= ( [A-Z] | '_' [a-zA-Z0-9_] ) [a-zA-Z0-9_]*
DOCCOMMENT?
         ::= '///' .* #xA
LINE_COMMENT?
         ::= '//' .* #xA
COMMENT? ::= '/*' .* '*/'
WS       ::= [ #x9#xd#xa]+
EOF      ::= $
```

**Pull request closed without merging**